### PR TITLE
Issue/#682

### LINKED ISSUE #682: Asset management: Unable to delete images uploaded in error (M - v0.2)
Suggested fix: add a delete button below the preview.

### PR DESCRIPTION
**Pull request by darylhedley**
fixes #682 - enables soft delete of assets ready for some sort of cron clean up

### CODE REVIEW

**Comment by darylhedley** [comment]
Also adds deletedAt and deletedBy so we can clean up assets later. For now a soft delete is perfect!

File: lib/assetmanager.js [hunk 1]
```diff
@@ -198,6 +198,9 @@ exports = module.exports = {
     } else {
       options.populate = _.extend(pop, options.populate);
     }
+
+    // Extend the search by _isDeleted to not return the soft deleted assets
+    search = _.extend(search, {_isDeleted: false});
```
Comment (dennis-learningpool) [inline lib/assetmanager.js]: can we check search first to check if `isDeleted` is already set, and only default to false if not? That way, the api can still be used to fetch deleted assets if, for example, we want to add a gui to inspect assets that have been deleted.

File: lib/assetmanager.js [hunk 2]
```diff
@@ -799,14 +842,22 @@ exports = module.exports = {
...
+    this.trashAsset(req.params.id, function(error) {
       if (error) {
         return next(error);
       }
 
       res.statusCode = 200;
       return res.json({success: true});
-    });
+    })
+    /*this.destroyAsset(req.params.id, function (error) {
```
Comment (dennis-learningpool) [inline lib/assetmanager.js]: you can just delete this commented code

File: routes/lang/en-application.json [hunk 3]
```diff
@@ -158,6 +158,7 @@
 	"app.upload": "Upload",
 	"app.assetfilters": "Filters",
 	"app.assetcollections": "Collections",
+	"app.assetconfirmdelete": "Are you sure you want to delete this asset? This cannot be undone",
```
Comment (dennis-learningpool) [inline routes/lang/en-application.json]: "This cannot be undone" is kinda fibbing?

Comment (darylhedley) [inline routes/lang/en-application.json]: To the end user it cannot be undone..but will remove it

File: lib/assetmanager.js (hunk 2 above)
Comment (darylhedley) [inline lib/assetmanager.js]: Was thinking about leaving it as this seems to be a temporary solution

File: lib/assetmanager.js (hunk 1 above)
Comment (taylortom) [inline lib/assetmanager.js]: +1 for this, a 'recycle bin'-type plugin could be useful

**Comment by taylortom** [comment]
This is only a part-solution for #682. Is this what we want? 

We're just telling the user that they've deleted an asset when in actuality it's still there, we're just hiding it from them.

(I really like the soft delete feature, just seems a bit pointless without a 'hard' delete).

**Comment by brian-learningpool** [comment]
I agree that we do need a 'hard' delete as well, but deleting assets completely from disk is a much more involved change.  Before doing so you would have to confirm that the asset isn't used in any course, because if it is and it's deleted, and the course is published again, it will have broken links.

I think something like the soft delete approach in this PR followed by a cron which checks if any 'soft' deleted assets not in use are physically deleted from disk makes sense.  I'm happy enough for that to wait to the next version though.

**Comment by darylhedley** [comment]
Just pushed a bit of code to allow a trash bin approach for the frontend

**Comment by brian-learningpool** [comment]
There's a slight UI issue as in if you delete an asset, it's still visible in the preview pane.  Then if you click Edit you get the Loading screen... and the page errors.  I think on successful deletion of the asset the preview pane should be removed.

**Comment by taylortom** [comment]
Works well apart from issue mentioned by @brian-learningpool.

**Comment by darylhedley** [comment]
@taylortom @brian-learningpool fixed that issue - could you test it again please?

**Comment by taylortom** [comment]
Works for me, +1

**Comment by brian-learningpool** [comment]
+1

**Pull request merged**